### PR TITLE
[baremetal] Add double pk verification and double master secret calculation

### PR DESCRIPTION
**Pull request by AndrzejKurek**

### CODE REVIEW

File: library/pk.c [hunk 1]
```diff
@@ -1569,6 +1569,8 @@ int mbedtls_pk_verify_restartable( mbedtls_pk_context *ctx,
     if( verify_ret == 0 )
     {
         mbedtls_platform_random_delay();
+        verify_ret = pk_info_verify_func( MBEDTLS_PK_CTX_INFO( ctx ),
```
Comment (shelib01) [inline library/pk.c:1572]: @AndrzejKurek , it is better to call twice the function from highest application level, in this case from ssl level. Is it possible? Why was the reason to call from this specific place ?

Comment (AndrzejKurek) [reply library/pk.c:1572]: I assumed that the fact that we are performing the verification twice is more of an implementation than the application choice, but you're right that current implementation will perform all verification twice (so, for example, the certificate will also be verified twice). 
I can move it to ssl_cli.c, lines ~3096, if that's the only thing we want: https://github.com/ARMmbed/mbedtls/blob/858e4325d2eb55274950e9335d0c08a1326069c2/library/ssl_cli.c#L3096

Comment (shelib01) [reply library/pk.c:1572]: there is also server side verify should be taken into account
https://github.com/ARMmbed/mbedtls/blob/858e4325d2eb55274950e9335d0c08a1326069c2/library/ssl_srv.c#L4645

Comment (shelib01) [reply library/pk.c:1572]: But generally, we want only from ssl level, these to places only should be protected. We also should verify that the operation itself was not skipped (but without adding code size  :) :) :) )

Comment (AndrzejKurek) [reply library/pk.c:1572]: Do you have any solution in mind? Isn't protecting the verification implementations enough (They'll return `PLATFORM_FAULT` if they find anything problematic)?

File: library/ssl_tls.c [hunk 2]
```diff
@@ -2852,6 +2893,8 @@ int mbedtls_ssl_encrypt_buf( mbedtls_ssl_context *ssl,
             return( ret );
         }
 
+        encryption_status = ENCRYPTION_SUCCESS;
```
Comment (shelib01) [inline library/ssl_tls.c:2896]: @AndrzejKurek , may be better to implement it this way ?
encryption_status = mbedtls_cipher_crypt( &transform->cipher_ctx_enc,
                                   transform->iv_enc,	                                   transform->iv_enc,
                                   transform->ivlen,	                                   transform->ivlen,
                                   data, rec->data_len,	                                   data, rec->data_len,
                                   data, &olen ) ) != 0 )	                                   data, &olen ) ) != 0 );

this decrease the chance an attacker can skip   mbedtls_cipher_crypt() and jump directly to 
 encryption_status = ENCRYPTION_SUCCESS; line. 
I understand that you wanted to prevent encryption_status to be 0 as success value, but it seems that skip still possible this way

File: library/ssl_tls.c [hunk 3]
```diff
@@ -1887,16 +1878,60 @@ static int ssl_compute_master( mbedtls_ssl_handshake_params *handshake,
...
+ *
+ * Parameters:
+ * [in/out] handshake
+ *          [in] resume, premaster, extended_ms, calc_verify, tls_prf
+ *          [out] premaster (cleared)
+ * [out] master
+ * [in] ssl: optionally used for debugging and calc_verify
+ */
+static int ssl_compute_master( mbedtls_ssl_handshake_params *handshake,
+                               unsigned char *master,
```
Comment (shelib01) [inline library/ssl_tls.c:1896]: @AndrzejKurek , I thought a lot about this feature. I think there is no chance that some user data will be leaked or session will continue if shared secret was not calculated or calculated wrong. So  we can skip this double master secret calculation, it is not adding security there

File: library/ssl_tls.c [hunk 4]
```diff
@@ -2500,6 +2535,9 @@ static void ssl_extract_add_data_from_record( unsigned char* add_data,
     }
 }
 
+#define ENCRYPTION_SUCCESS 0xCC
+#define ENCRYPTION_FAIL 0xAA
```
Comment (shelib01) [inline library/ssl_tls.c:2539]: @AndrzejKurek , does it add us some security to use predefined values instead of regular error codes in the function?
Why we don't use  
int encryption_status = BEDTLS_ERR_PLATFORM_FAULT_DETECTED;

File: library/ssl_srv.c [hunk 5]
```diff
@@ -4643,13 +4643,16 @@ static int ssl_parse_certificate_verify( mbedtls_ssl_context *ssl )
     }
 
     ret = mbedtls_pk_verify( peer_pk,
-                                      md_alg, hash_start, hashlen,
-                                      ssl->in_msg + i, sig_len );
+                             md_alg, hash_start, hashlen,
```
Comment (shelib01) [inline library/ssl_srv.c:4646]: @AndrzejKurek , why ther return code in this function is initialized this way?
https://github.com/ARMmbed/mbedtls/blob/7f656459f773bd445d1616e90ed420a5c9210973/library/ssl_srv.c#L4459

Do we want to add additional error code for verify function to prevent its' skipping ?

File: library/ssl_cli.c [hunk 6]
```diff
@@ -3100,6 +3100,11 @@ static int ssl_in_server_key_exchange_parse( mbedtls_ssl_context *ssl,
         {
             mbedtls_platform_random_delay();
 
+            if( rs_ctx == NULL )
```
Comment (shelib01) [inline library/ssl_cli.c:3103]: @AndrzejKurek , could you explain this restartable context check?

Comment (shelib01) [reply library/ssl_cli.c:3103]: Do we need special return value for this function to prevent skipping ?

File: library/ssl_srv.c (hunk 5 above)
Comment (AndrzejKurek) [reply library/ssl_srv.c:4646]: Good question. I think that we can change it to the platform fault error and test if CI passes, I don't see a use of this value possible in the function flow.

File: library/ssl_cli.c (hunk 6 above)
Comment (AndrzejKurek) [reply library/ssl_cli.c:3103]: The thing is that if there's a restartable context set, verification can stop at any given point (because a maximal number of operations has been reached), and then, when it's called the next time, it will continue from this point.
So, now, in context of these double calls - we won't receive the same result with two subsequent calls.
As for the special return value - the checks here are so nested that I wouldn't expect a need to create an additional variable.

File: library/ssl_tls.c (hunk 4 above)
Comment (AndrzejKurek) [reply library/ssl_tls.c]: Added.

File: library/ssl_tls.c (hunk 3 above)
Comment (AndrzejKurek) [reply library/ssl_tls.c]: Removed!

File: library/ssl_tls.c (hunk 2 above)
Comment (AndrzejKurek) [reply library/ssl_tls.c]: Added.

**Comment by AndrzejKurek** [comment]
This PR adds 44 bytes to the codesize on gcc_arm.

**Review by shelib01 — DISMISSED** [review]

**Comment by AndrzejKurek** [comment]
Rebased on top of https://github.com/ARMmbed/mbedtls/pull/3724 to eliminate conflicts. The plan is to merge 3724 first and then have a clear history in this PR.

File: library/ssl_srv.c [hunk 7]
```diff
@@ -4643,13 +4643,16 @@ static int ssl_parse_certificate_verify( mbedtls_ssl_context *ssl )
...
+                             ssl->in_msg + i, sig_len );
 
     if( ret == 0 )
     {
         mbedtls_platform_random_delay();
 
+        ret = mbedtls_pk_verify( peer_pk,
```
Comment (shelib01) [inline library/ssl_srv.c:4653]: @AndrzejKurek I still feel that we have a weak point here.  Let's add additional return code for the second verify and return success only in case both return codes are true. In this case we will know that both were not skipped. The same on the client side.

Comment (AndrzejKurek) [reply library/ssl_srv.c]: Added!

File: library/ssl_cli.c [hunk 8]
```diff
@@ -2794,6 +2794,7 @@ static int ssl_in_server_key_exchange_parse( mbedtls_ssl_context *ssl,
      * structural change to provide default flow assumes failure
      */
     volatile int ret = 0;
+    volatile int ret_fi = MBEDTLS_ERR_PLATFORM_FAULT_DETECTED;
```
Comment (shelib01) [inline library/ssl_cli.c:2797]: @AndrzejKurek , why did you change  volatile `int ret = 0;` to be initialized to 0 instead of  MBEDTLS_ERR_PLATFORM_FAULT_DETECTED ? I am almost sure it was not zero in previous version

Comment (AndrzejKurek) [reply library/ssl_cli.c:2797]: It is and was zero - there's a comment added just above it - changing it to a different value would require significant rework of this function, and this is an even bigger risk.

Comment (shelib01) [reply library/ssl_cli.c:2797]: @AndrzejKurek , OK, I missed the comment. Lets commit it with 0, but i think we should investigate this function and change its' flow in the future. I don't like that function can just jump to the end, and return success :(

**Comment by shelib01** [comment]
@AndrzejKurek , why tests do not run ?

**Review by shelib01 — APPROVED** [review]

**Comment by AndrzejKurek** [comment]
> @AndrzejKurek , why tests do not run ?

We will probably need to wait, as the CI is overloaded.

**Comment by AndrzejKurek** [comment]
It looks like `git fetch` failed on one of the machines, all other tests are green. Should I re-run the test, just to be sure, or should we merge? @shelib01
